### PR TITLE
ACS-3198: Folder rules v1 REST API - List rules for node

### PR DESCRIPTION
**Pull request by krdabrowski**
Get folder rules API: `GET /nodes/{folderNodeId}/rulesets/{ruleSetId}/rules` implementation proposal.

### CODE REVIEW

File: remote-api/src/main/java/org/alfresco/rest/api/Rules.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/Rules.java:38]: Add @Experimental annotation for now. Similar for other new interfaces & classes ?

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java:39]: (Why) do we need to add this subclass to the REST framework ?

Is this for in-memory paging (if underlying service does not support paging - with assumption that items are listed in a repeatable order over multiple paged calls) ?

If so, I wonder if there are any other existing implemented areas that do something similar that might benefit ?

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (montgolfiere) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:60]: Does is account for sub-types of cm:folder (it might, I haven't checked yet) ?

File: remote-api/src/main/java/org/alfresco/rest/api/Rules.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/Rules.java:38]: Ok, I will add annotations everywhere where it's needed.

Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/Rules.java:40]: Annotations added

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java
Comment (montgolfiere) [reply remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java:41]: Thanks. 

Do we have a list of "everyone" (eg. code links) that we could add to your comment (for reference) ?

Maybe also log as tech debt in Jira (ie. option to go back and rationalise others) to use your generic approach ?

**Review by montgolfiere — COMMENTED** [review]
LGTM

Before we merge, we should decide along with other tech leads (cc @tpage-alfresco):

1. Minor: maybe add a few more key fields, in particular, maybe description and enabled fag. I think these two fields are useful for at least the "linked rules" use-case, ie. UI picker to select a folder with existing rules to link to.

2. Major: maybe wait for API design decision as to whether we will change this endpoint to be one more level below, eg. GET /nodes/{folderNodeId}/rulesets/-default-/rules

Otherwise, we merge as experimental with the knowledge that we will update & maybe refactor soon with further PRs. 

Either, we should also ensure we keep in-sync with proposed API spec (and vice-versa).

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPages.java:41]: "Everyone" is not a proper wording. A lot of solutions (better sentence) handles the paging in its custom way. I don't have any list, but it's possible to do some research. Some of the solutions are old and exists there for long time.

**Comment by tpage-alfresco** [comment]
> For #2, I see @tpage-alfresco has now added extra level to the [PR](https://github.com/Alfresco/rest-api-explorer/pull/161) for the API spec.
> 
> It probably makes sense to change now so that we don't get too far out-of-sync, especially as we add more endpoints (eg. to get rule, create rule, update rule and delete rule) and start initial client dev work.

I'm still not sure which is best tbh.  I think we need to work on the Word doc a bit more to get all the interactions ironed out (including potential future ones). This should give us a better idea of what the endpoints should be.

**Review by montgolfiere — APPROVED** [review]
LGTM - added a minor comment about Auditable (also trivial / subjective one about code).

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:50]: At some point I think we'll want a Ruleset object and then this might want to move there.  For now this is great.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:65]: Does this end up as a 404 error?  I would assume we want 404 for this error anyway.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:69]: Maybe we should inline this variable too?  We don't use the rulesModels variable except to immediately stream it.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:85]: Please could we have some javadoc to say what counts as validation for a node?

File: remote-api/src/main/java/org/alfresco/rest/api/nodes/NodeRuleSetsRelation.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/api/nodes/NodeRuleSetsRelation.java:38]: Ugh! I thought we used the term "ruleset", but Googling suggests we call it a "rule set."  I think we should change the URL too.

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java:52]: I don't understand why this returns a pair.

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java
Comment (tpage-alfresco) [inline remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java:74]: I don't understand what it means to create a ListPage from no list.

File: repository/src/main/java/org/alfresco/service/cmr/rule/RuleService.java
Comment (krdabrowski) [reply repository/src/main/java/org/alfresco/service/cmr/rule/RuleService.java:341]: Fixed

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:69]: Good point! I'll fix it.

File: remote-api/src/main/java/org/alfresco/rest/api/nodes/NodeRuleSetsRelation.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/nodes/NodeRuleSetsRelation.java:38]: I'll fix it.

File: remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/framework/resource/parameters/ListPage.java:52]: This is a contract method from `interface PagingResults<R>`. Can't really do much about it.

File: remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java
Comment (krdabrowski) [reply remote-api/src/main/java/org/alfresco/rest/api/impl/RulesImpl.java:91]: On it.

**Pull request merged**
